File: src/wizctl/genmon.py
"""XFCE Generic Monitor (xfce4-genmon-plugin) status provider."""

import os
from pathlib import Path
import sys
from typing import Optional

from pywizlight import SCENES

from wizctl.bulb import DEFAULT_BULB_IP
from wizctl.state import load_state


def get_panel_asset_dir() -> Path:
    """Ensure panel icons are saved in a persistent user directory for XFCE genmon."""
    user_assets = Path.home() / ".local" / "share" / "wizctl" / "assets"
    user_assets.mkdir(parents=True, exist_ok=True)
    
    bundled = Path(__file__).parent / "assets"
    if bundled.is_dir():
        for f in bundled.glob("*.png"):
            target = user_assets / f.name
            if not target.exists() or target.stat().st_size != f.stat().st_size:
                try:
                    target.write_bytes(f.read_bytes())
                except Exception:
                    pass
    return user_assets


def generate_genmon_xml(target_ip: Optional[str] = None) -> str:
    """Generate XML output conforming to xfce4-genmon-plugin specification."""
    state = load_state()
    ip = target_ip or state.get("ip", DEFAULT_BULB_IP)

    assets_dir = get_panel_asset_dir()
    
    power = state.get("power", True)
    brightness = state.get("brightness", 255)
    pct = int(brightness * 100 / 255)
    scene_id = state.get("scene_id")
    mode = state.get("mode", "color")
    kelvin = state.get("kelvin", 2700)
    hex_code = state.get("hex", "#ff8c00")

    # Icon selection
    if power:
        icon_path = assets_dir / "panel_bulb_on.png"
        status_text = f"{pct}%"
        if mode == "scene" and scene_id:
            sname = SCENES.get(scene_id, f"Scene {scene_id}")
            status_text = f"{sname} ({pct}%)"
        power_str = "ON"
    else:
        icon_path = assets_dir / "panel_bulb_off.png"
        status_text = "OFF"
        power_str = "OFF"

    # Tooltip
    tooltip_lines = [
        f"WiZ Smart Light ({ip})",
        f"Status: {power_str}",
        f"Brightness: {pct}% ({brightness}/255)",
    ]
    if mode == "scene" and scene_id:
        tooltip_lines.append(f"Scene: {SCENES.get(scene_id, f'ID {scene_id}')}")
    elif mode == "kelvin":
        tooltip_lines.append(f"White Temp: {kelvin}K")
    else:
        tooltip_lines.append(f"Color: {hex_code}")
    tooltip_lines.append("Left-click: Quick Widget | Double-click: Toggle")
    tooltip = "\n".join(tooltip_lines)

    # Click command: uses our double-click/single-click handler
    # Locate wizctl binary or python script
    click_cmd = "wizctl widget --click"

    xml = (
        f"<img>{icon_path.resolve()}</img>\n"
        f"<txt> {status_text} </txt>\n"
        f"<tool>{tooltip}</tool>\n"
        f"<click>{click_cmd}</click>\n"
    )
    return xml


def run_genmon(target_ip: Optional[str] = None) -> int:
    """CLI runner for genmon output."""
    sys.stdout.write(generate_genmon_xml(target_ip=target_ip))
    return 0

File: src/genmon.rs
use std::env;
use std::fs;
use std::path::PathBuf;

use crate::colors::get_scene_name;
use crate::state::load_state;

const ICON_ON_BYTES: &[u8] = include_bytes!("../assets/panel_bulb_on.png");
const ICON_OFF_BYTES: &[u8] = include_bytes!("../assets/panel_bulb_off.png");
const ICON_OFFLINE_BYTES: &[u8] = include_bytes!("../assets/panel_bulb_offline.png");

pub fn get_panel_asset_dir() -> PathBuf {
    let base_dir = if let Ok(data_home) = env::var("XDG_DATA_HOME") {
        PathBuf::from(data_home).join("wizctl").join("assets")
    } else if let Ok(home) = env::var("HOME") {
        PathBuf::from(home).join(".local").join("share").join("wizctl").join("assets")
    } else {
        PathBuf::from("/tmp/wizctl/assets")
    };

    let _ = fs::create_dir_all(&base_dir);

    // Auto-extract embedded icons if missing or empty
    let on_path = base_dir.join("panel_bulb_on.png");
    if !on_path.exists() || fs::metadata(&on_path).map(|m| m.len()).unwrap_or(0) == 0 {
        let _ = fs::write(&on_path, ICON_ON_BYTES);
    }

    let off_path = base_dir.join("panel_bulb_off.png");
    if !off_path.exists() || fs::metadata(&off_path).map(|m| m.len()).unwrap_or(0) == 0 {
        let _ = fs::write(&off_path, ICON_OFF_BYTES);
    }

    let offline_path = base_dir.join("panel_bulb_offline.png");
    if !offline_path.exists() || fs::metadata(&offline_path).map(|m| m.len()).unwrap_or(0) == 0 {
        let _ = fs::write(&offline_path, ICON_OFFLINE_BYTES);
    }

    base_dir
}

pub fn generate_genmon_xml(target_ip: Option<&str>) -> String {
    let state = load_state();
    let ip = target_ip.unwrap_or(&state.ip);
    let assets_dir = get_panel_asset_dir();

    let pct = (state.brightness as f64 * 100.0 / 255.0).round() as u8;
    let power_str = if state.power { "ON" } else { "OFF" };

    let (icon_path, status_text) = if state.power {
        let text = if state.mode == "scene" && state.scene_id > 0 {
            let sname = get_scene_name(state.scene_id).unwrap_or("Scene");
            format!("{sname} ({pct}%)")
        } else {
            format!("{pct}%")
        };
        (assets_dir.join("panel_bulb_on.png"), text)
    } else {
        (assets_dir.join("panel_bulb_off.png"), "OFF".to_string())
    };

    let mut tooltip_lines = vec![
        format!("WiZ Smart Light ({ip})"),
        format!("Status: {power_str}"),
        format!("Brightness: {pct}% ({}/255)", state.brightness),
    ];

    if state.mode == "scene" && state.scene_id > 0 {
        let sname = get_scene_name(state.scene_id).unwrap_or("Unknown");
        tooltip_lines.push(format!("Scene: {sname} (ID {})", state.scene_id));
    } else if state.mode == "kelvin" && state.kelvin > 0 {
        tooltip_lines.push(format!("White Temp: {}K", state.kelvin));
    } else {
        tooltip_lines.push(format!("Color: {}", state.hex));
    }

    tooltip_lines.push("Left-click: Quick Widget | Double-click: Toggle".to_string());
    let tooltip = tooltip_lines.join("\n");

    let click_cmd = "wizctl widget --click";

    format!(
        "<img>{}</img>\n<txt> {} </txt>\n<tool>{}</tool>\n<click>{}</click>\n",
        icon_path.display(),
        status_text,
        tooltip,
        click_cmd
    )
}

pub fn run_genmon(target_ip: Option<&str>) -> Result<(), Box<dyn std::error::Error>> {
    print!("{}", generate_genmon_xml(target_ip));
    Ok(())
}
